crates/turborepo-filewatch/src/optional_watch.rs

```diff
@@ -0,0 +1,68 @@
+use tokio::sync::watch::{self, error::RecvError, Ref};
+
+#[derive(Debug)]
+pub struct OptionalWatch<T>(watch::Receiver<Option<T>>);
+
+impl<T> Clone for OptionalWatch<T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+/// A handy wrapper around types that are watched and may be None.
+/// `SomeRef` is a reference type that is known to be `Some`.
+impl<T> OptionalWatch<T> {
+    /// Create a new `OptionalWatch` with no initial value.
+    ///
+    /// Keep in mind that when the sender is dropped, down stream
+    /// dependencies will get a RecvError.
+    pub fn new() -> (watch::Sender<Option<T>>, OptionalWatch<T>) {
+        let (tx, rx) = watch::channel(None);
+        (tx, OptionalWatch(rx))
+    }
+
+    /// Create a new `OptionalWatch` with an initial, unchanging value.
+    pub fn once(init: T) -> Self {
+        let (_tx, rx) = watch::channel(Some(init));
+        OptionalWatch(rx)
+    }
+
+    /// Wait for the value to be available and then return it.
+    ///
+    /// If you receive a `RecvError`, the sender has been dropped, meaning you
+    /// will not receive any more updates. For efficiencies sake, you should
+    /// exit the task and drop any senders to other dependencies so that the
+    /// exit can propagate up the chain.
+    pub async fn get(&mut self) -> Result<SomeRef<'_, T>, RecvError> {
```

Comment:
I'm a little confused by `SomeRef`. Could we unwrap here and return `&'a T`? Since we can turn `Ref<'a, Option<T>>` into a `&'a Option<T>` and then into `Option<&'a T>` which we unwrap. Also it might be worth noting that any ref here blocks the producer

Reply:
I had a look here and as I understand it, a Ref is needed as it unlocks the watcher when it is dropped, similar to a MutexGuard. We could try to use `as_ref().unwrap()` but the lifetime of `&T` is tied to the `Ref` inside the function, not to the function itself.